### PR TITLE
Cart life cycle

### PR DESCRIPTION
**Pull request by MatteoPiovanelli-Laser**
part of the things discussed in #117 

In this PR I am adding cart life cycle events. I made an event interface for the events on a cart I could think of:

- adding an item
- removing an item
- adding/removing in bulk
- "finalization" of the cart, meaning the customer bought what was in it.
- update, generic, called by all the others except `Finalized`.

I provided a default implementation for the event handler, that will trigger wrokflow activities specific for all steps.

I also made a few changes to the tokens for the cart (and added one for the item added/removed), because it wass not outputing the strings for products as I expected them to (attributes where not written out, instead having their class name as a string).

### CODE REVIEW

File: Controllers/ShoppingCartController.cs [hunk 1]
```diff
@@ -382,10 +379,21 @@ public class ShoppingCartController : Controller {
...
 
         private void UpdateShoppingCart(IEnumerable<UpdateShoppingCartItemViewModel> items) {
+            var oldItems = new List<ShoppingCartItem>();
+            oldItems.AddRange(_shoppingCart.Items); //create a copy for analysis
             _shoppingCart.Clear();
 
-            if (items == null)
+            if (items == null) {
+                //removed all items
+                foreach (var han in _cartLifeCycleEventHandlers) {
```
Comment (bleroy) [inline Controllers/ShoppingCartController.cs]: We can write `handler` in its entirety ;)

File: Models/ProductAttributeValueExtended.cs [hunk 2]
```diff
@@ -21,5 +21,9 @@ public class ProductAttributeValueExtended : IEquatable<ProductAttributeValueExt
                 this.ExtendedValue == other.ExtendedValue &&
                 this.ExtensionProvider == other.ExtensionProvider;
         }
+
+        public override string ToString() {
+            return "(" + string.Join(", ", new string[] { Value, ExtendedValue, ExtensionProvider }) + ")";
```
Comment (bleroy) [inline Models/ProductAttributeValueExtended.cs]: `$"({Value}, {ExtendedValue}, {ExtensionProvider})"`

File: Controllers/ShoppingCartController.cs (hunk 1 above)
Comment (MatteoPiovanelli-Laser) [inline Controllers/ShoppingCartController.cs]: I am coding on a Lettera 32, and writing han saves me ink.
(I changed this to `handler` through all the code)

File: Models/ProductAttributeValueExtended.cs (hunk 2 above)
Comment (MatteoPiovanelli-Laser) [inline Models/ProductAttributeValueExtended.cs]: Yet another language feature I always forget it's there.

**Comment by MatteoPiovanelli-Laser** [comment]
Updated the PR

**Comment by bleroy** [comment]
Thanks!

**Pull request merged**